makeflow/src/makeflow_module_resource_monitor.c

```diff
@@ -0,0 +1,345 @@
+
+#include "create_dir.h"
+#include "debug.h"
+#include "jx.h"
+#include "path.h"
+#include "rmonitor.h"
+#include "stringtools.h"
+#include "xxmalloc.h"
+
+#include "batch_task.h"
+#include "batch_wrapper.h"
+
+#include "dag.h"
+#include "dag_file.h"
+#include "dag_node.h"
+#include "makeflow_hook.h"
+#include "makeflow_log.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define DEFAULT_MONITOR_LOG_FORMAT "resource-rule-%%"
+
+struct makeflow_monitor {
+	int enable_debug;
+	int enable_time_series;
+	int enable_list_files;
+
+	int interval;
+	char *log_dir;
+	char *log_format;
+	char *log_prefix;
+	char *exe;
+	const char *exe_remote;
+};
+
+struct makeflow_monitor *monitor = NULL;
+
+struct makeflow_monitor *makeflow_monitor_create()
+{
+	struct makeflow_monitor *m = malloc(sizeof(*m));
+	m->enable_debug = 0;
+	m->enable_time_series = 0;
+	m->enable_list_files = 0;
+
+	m->interval = 1; // in seconds
+	m->log_dir = NULL;
+	m->log_format = NULL;
+	m->log_prefix = NULL;
+	m->exe = NULL;
+	m->exe_remote = NULL;
+
+	return m;
+}
+
+static int create(struct jx *args)
+{
+	monitor = makeflow_monitor_create();
+	if (jx_lookup_string(args, "resource_monitor_log_dir"))
+		monitor->log_dir = xxstrdup(jx_lookup_string(args, "resource_monitor_log_dir"));
+
+	if (jx_lookup_string(args, "resource_monitor_log_format"))
+		monitor->log_format = xxstrdup(jx_lookup_string(args, "resource_monitor_log_format"));
+
+	if(jx_lookup_integer(args, "resource_monitor_interval"))
+		monitor->interval = jx_lookup_integer(args, "resource_monitor_interval");
+
+	monitor->enable_time_series = jx_lookup_integer(args, "resource_monitor_enable_time_series");
+	monitor->enable_list_files = jx_lookup_integer(args, "resource_monitor_enable_list_files");
+
+	if (!monitor->log_dir) {
+		fatal("Monitor mode was enabled, but a log output directory was not specified (use --monitor=<dir>)");
+		return MAKEFLOW_HOOK_FAILURE;
+	}
+
+	if (!monitor->log_format)
+		monitor->log_format = xxstrdup(DEFAULT_MONITOR_LOG_FORMAT);
+
+	monitor->log_prefix = string_format("%s/%s", monitor->log_dir, monitor->log_format);
+
+	if (monitor->interval < 1) {
+		fatal("Monitoring interval should be positive.");
+		return MAKEFLOW_HOOK_FAILURE;
+	}
+
+	monitor->exe = resource_monitor_locate(NULL);
+	if (!monitor->exe) {
+		fatal("Monitor mode was enabled, but could not find resource_monitor in PATH.");
+	}
+
+	monitor->exe_remote = path_basename(monitor->exe);
```

Comment:
Won't both of these resolve to the same thing? I used path_basename in the off chance that the executable name of the resource_monitor may change. I was hoping to avoid names where possible.

Reply:
The thinking when doing this in work_queue is that resource_monitor may be a common name used by the task, while cctools-monitor is something clearly ours.

Ideally, it should be a name known not to exists in the task, but that needs more machinery, and I don't think we want to spend time in that.

Reply:
Ok. I can agree with that. I will change this.